I/5762: Restricted editing mode plugin - restricing editing in restricted editing mode

jodator opened the pull request:
### Suggested merge commit message ([convention](https://github.com/ckeditor/ckeditor5-design/wiki/Git-commit-message-convention))

Feature: Enforcing restricted editing mode. Closes ckeditor/ckeditor5#5762.

---

### Additional information

This PR contains:

- enforcing restricted editing mode by toggling editor commands - allowed commands are configurable
- marker editing enhancements (typing with a marker)
- class names refactoring (`RestrictedEditingMode` and `StandardEditingMode` plugins)
- stub of a feature guide - to enable testing phase
- fixed collapsed marker conversion (additional converter as `markerToHighlight` and markers have their own issues)
- pasting is blocked in restricted editing mode (should be enabled in a follow up)

----

Requires https://github.com/ckeditor/ckeditor5-theme-lark/pull/255

Reinmar commented on src/restrictededitingmodeediting.js:
```diff
@@ -0,0 +1,275 @@
+		/**
+		 * Command names that are enabled outside non-restricted regions.
+		 *
+		 * @type {Set<string>}
```
```suggestion
		 * @type {Set.<String>}
```

Reinmar commented on src/restrictededitingmodeediting.js:
```diff
@@ -0,0 +1,275 @@
+		super( editor );
+
+		editor.config.define( 'restrictedEditing', {
+			allowedCommands: [ 'bold', 'italic', 'link' ]
```
```suggestion
			allowedCommands: [ 'bold', 'italic', 'link', 'unlink' ]
```

Reinmar commented on src/restrictededitingmodeediting.js:
```diff
@@ -0,0 +1,275 @@
+		// That's why there are 2 downcast converters for them:
+		// 1. The default marker-to-highlight will wrap selected text with `<span>`.
+		editor.conversion.for( 'downcast' ).markerToHighlight( {
+			model: 'restricted-editing-exception',
```
`restrictedEditingException` since this is a code name. We use the dashed convention for file names, ids and CSS classes (due to BEM).

Reinmar commented on src/restrictededitingmodeediting.js:
```diff
@@ -0,0 +1,275 @@
+		this._setupCommandsToggling();
+
+		// Commands & keystrokes that allow navigation in the content.
+		editor.commands.add( 'goToPreviousRestrictedEditingRegion', new RestrictedEditingNavigationCommand( editor, 'backward' ) );
```
`goToPreviousRestrictedEditingException`

You use the `restricted-editing-exception` name in other places already and it should be fine.

Reinmar commented on src/restrictededitingmodeediting.js:
```diff
@@ -0,0 +1,275 @@
+		editor.conversion.for( 'downcast' ).markerToHighlight( {
+			model: 'restricted-editing-exception',
+			// Use callback to return new object every time new marker instance is created - otherwise it will be seen as the same marker.
+			view: () => ( {
```
TBH, I prefer:

```js
view() {
     return { 
         foo: 1,
         bar: 2
     };
}
```

Not a big deal but I think that great majority of currently existing code follows that. There are only 7 occurrences of `() => ( {`.

Reinmar replied:
But we can ask more people for their opinions.

Note: below there's another fragment that I'd rather change:

```js
editor.conversion.for( 'dataDowncast' ).markerToElement( {
	model: 'restricted-editing-exception',
	view: ( markerData, viewWriter ) => viewWriter.createEmptyElement( 'span', {
		class: 'ck-restricted-editing-exception'
	} )
} );
```

jodator commented:
> ## Bugs
> 
>     * Pasting doesn't work.

I've decided that it is too much for the MVP and could be postponned: https://github.com/ckeditor/ckeditor5/issues/5802

>     * You can delete text by backspace at the beginning of an editable region. Same for fwdDelete at the end.

I was refactoring the code and it turned out that this case wasn't tested properly (I've removed the code and the test passed). Fixed

>     * Composition at the end of a editable region doesn´t work.

I'd go with a follow-up : https://github.com/ckeditor/ckeditor5/issues/5827

>     * Ctrl/Cmd+X isn't blocked in non-editable regions.

Added

## Improvements
* Removing exceptions via collapsed selection should remove the entire exception (like in the highlight feature).
* Editable regions should react to hover in some way (border, color, cursor?): https://github.com/ckeditor/ckeditor5/issues/5801
* Ctrl/Cmd+A should select only the content of an editable region. https://github.com/ckeditor/ckeditor5/issues/5826
* Empty region is bigger than a region with one letter. It could be smaller I guess. (moved that also to https://github.com/ckeditor/ckeditor5/issues/5801)
